src/libraries/System.Private.CoreLib/src/System/IO/FileSystemInfo.Unix.cs

```diff
@@ -69,7 +69,7 @@ public void Refresh()
             _fileStatus.RefreshCaches(FullPath);
         }
 
-        internal static void ThrowNotFound(string path)
+        internal static void ThrowNotFound(ReadOnlySpan<char> path)
```

Comment:
In some cases we're switching from `string` to `ROS` and back to `string`. Can we accept `string?`.
```

        internal static void ThrowNotFound(string? path)
        {
            // Windows distinguishes between whether the directory or the file isn't found,
            // and throws a different exception in these cases.  We attempt to approximate that
            // here; there is a race condition here, where something could change between
            // when the error occurs and our checks, but it's the best we can do, and the
            // worst case in such a race condition (which could occur if the file system is
            // being manipulated concurrently with these checks) is that we throw a
            // FileNotFoundException instead of DirectoryNotFoundException.
            bool directoryError = path is not null && !Directory.Exists(Path.GetDirectoryName(Path.TrimEndingDirectorySeparator(path)));
            throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(Interop.Error.ENOENT), path, directoryError);
        }
```

Reply:
Also, I think you should be able to replace `Directory.Exists` with `FileSystem.DirectoryExists`, and use `path.AsSpan()` as the argument for `TrimEndingDirectorySeparator`.